[nanobind] New port

ekilmer opened the pull request:
Nanobind distributes its source files to build on-demand, so I have added a CI port to make sure this continues to work with the [provided upstream example](https://github.com/wjakob/nanobind_example) by the author of nanobind.

After testing, cross-compilation for a different architecture (arm64-windows in CI) is not supported due to the execution of the Python executable to figure out the Python module extension.

Closes: #33401 

Supersedes: #31323 


- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

ekilmer committed:
[nanobind] Add new port

Nanobind distributes its source files and CMake helper functions to use
the project.

Nanobind does not support cross-compilation for different architectures.

ekilmer commented:
I'm very confused about what's happening. Why is `python311.lib` being listed for the Debug builds? And why does this only happen on Windows dynamic triplets? It's not listed in the link command (`python311_d.lib` is).

Snippet:
```
[13/13] cmd.exe /C "cd . && D:\downloads\tools\cmake-3.27.1-windows\cmake-3.27.1-windows-i386\bin\cmake.exe -E vs_link_dll --intdir=CMakeFiles\nanobind_example_ext.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x64\link.exe  CMakeFiles\nanobind_example_ext.dir\src\nanobind_example_ext.cpp.obj  /out:nanobind_example_ext.cp311-win_amd64.pyd /implib:nanobind_example_ext.lib /pdb:nanobind_example_ext.pdb /dll /version:0.0 /machine:x64 /nologo    /debug /INCREMENTAL  nanobind-static.lib  D:\installed\x64-windows\debug\lib\python311_d.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
FAILED: nanobind_example_ext.cp311-win_amd64.pyd 
cmd.exe /C "cd . && D:\downloads\tools\cmake-3.27.1-windows\cmake-3.27.1-windows-i386\bin\cmake.exe -E vs_link_dll --intdir=CMakeFiles\nanobind_example_ext.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x64\link.exe  CMakeFiles\nanobind_example_ext.dir\src\nanobind_example_ext.cpp.obj  /out:nanobind_example_ext.cp311-win_amd64.pyd /implib:nanobind_example_ext.lib /pdb:nanobind_example_ext.pdb /dll /version:0.0 /machine:x64 /nologo    /debug /INCREMENTAL  nanobind-static.lib  D:\installed\x64-windows\debug\lib\python311_d.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib  && cd ."
LINK Pass 1: command "C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1437~1.328\bin\Hostx64\x64\link.exe CMakeFiles\nanobind_example_ext.dir\src\nanobind_example_ext.cpp.obj /out:nanobind_example_ext.cp311-win_amd64.pyd /implib:nanobind_example_ext.lib /pdb:nanobind_example_ext.pdb /dll /version:0.0 /machine:x64 /nologo /debug /INCREMENTAL nanobind-static.lib D:\installed\x64-windows\debug\lib\python311_d.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:CMakeFiles\nanobind_example_ext.dir/intermediate.manifest CMakeFiles\nanobind_example_ext.dir/manifest.res" failed (exit code 1104) with the following output:
LINK : fatal error LNK1104: cannot open file 'python311.lib'
```

ekilmer commented:
After running the `link.exe` command locally with `/VERBOSE`, I see `python311.lib` mentioned at the top of the output and then not again until it fails to find it:

```
Starting pass 1
Processed /DEFAULTLIB:msvcprtd
Processed /DEFAULTLIB:python311.lib
Processed /DEFAULTLIB:MSVCRTD
Processed /DEFAULTLIB:OLDNAMES
```

Searching for `DEFAULTLIB` doesn't yield any obvious locations that would cause `python311.lib` to appear here. I searched the source code of nanobind, CMake, vcpkg, and nanobind_example.

After some searching, I think this same issue (https://github.com/microsoft/vcpkg/issues/23052#issuecomment-1345600775) is present in nanobind: https://github.com/wjakob/nanobind/blob/eb308de0e3099e5abd8e2e9b60a8e0af81ee93db/include/nanobind/nb_python.h#L11-L19

ekilmer commented:
This likely would have worked if the debug headers were not removed because `pyconfig.h` sets `Py_DEBUG 1` for debug builds, and nanobind respects that variable.

LilyWangLL commented:
Convert this PR to draft since there are errors on CI pipeline need to fix. Please ping us if this PR is ready for review again.

dg0yt commented on scripts/test_ports/vcpkg-ci-nanobind/portfile.cmake:
```diff
@@ -0,0 +1,23 @@
+
+# This is needed to correctly build/link against a debug build of Python on
+# Windows
+set(VCPKG_CXX_FLAGS_DEBUG -DPy_DEBUG ${VCPKG_CXX_FLAGS_DEBUG})
+set(VCPKG_C_FLAGS_DEBUG -DPy_DEBUG ${VCPKG_C_FLAGS_DEBUG})
```
Note that these are strings (as on the command line), not CMake lists.
```suggestion
string(APPEND VCPKG_C_FLAGS_DEBUG " -DPy_DEBUG")
string(APPEND VCPKG_CXX_FLAGS_DEBUG " -DPy_DEBUG")
```

ekilmer replied:
Thank you!

dg0yt commented on scripts/test_ports/vcpkg-ci-nanobind/portfile.cmake:
```diff
@@ -0,0 +1,23 @@
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DPython_EXECUTABLE=${PYTHON3}
```
```suggestion
        "-DPython_EXECUTABLE=${PYTHON3}"
```

ekilmer replied:
I removed this. I don't think this is correct for cross-compilation anyway. Nanobind tries to use the Python executable to learn details about extension suffix, but if it's using the host's Python, then this will be wrong for the target.

Probably just easiest to ignore cross-compilation for now.

dg0yt commented on ports/nanobind/vcpkg.json:
```diff
@@ -0,0 +1,18 @@
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
```
This port provides `vcpkg_cmake_config_fixup`, but I don't see this function being used.

ekilmer replied:
Oops. Good catch. Thank you.

dg0yt commented on scripts/test_ports/vcpkg-ci-nanobind/portfile.cmake:
```diff
@@ -0,0 +1,23 @@
+set(VCPKG_CXX_FLAGS_DEBUG -DPy_DEBUG ${VCPKG_CXX_FLAGS_DEBUG})
+set(VCPKG_C_FLAGS_DEBUG -DPy_DEBUG ${VCPKG_C_FLAGS_DEBUG})
+
+vcpkg_find_acquire_program(PYTHON3)
```
AFAIK this provides the interpreter which runs on the host, but not the development stuff fort the target. The port probably needs the `python3` port, at least for the target triplet. But I cannot help with the details.

ekilmer replied:
Yeah, I don't think this is the solution for cross-compilation, so I've removed it. See the comment here for more thoughts https://github.com/microsoft/vcpkg/pull/35488#discussion_r1429005897

ekilmer commented:
CI is now passing. I am waiting for upstream to make a new release that includes the changes that make packaging easier.

I will leave this PR as a draft until nanobind makes a new release, but if there are any other comments on the content of this PR, I am happy to address them!

ADGrant commented:
@ekilmer   When do you expect nano bind to release a version with the expected changes for vcpkg compatibility?

ekilmer commented on ports/nanobind/vcpkg.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "nanobind",
+  "version-date": "2023-12-16",
```
Note: This should be changed to
```suggestion
  "version-semver": "1.9.0",
```
when the next version (likely `1.9.0`) is released.

I have chosen `version-semver` based on the text in nanobind's [changelog file](https://github.com/wjakob/nanobind/blob/16515dbe198d9406fad046d7fc6e7c4ca5e2c5a6/docs/changelog.rst)

ekilmer replied:
N/A. I've decided to backport the CMake patches required for packaging.

ekilmer commented on ports/nanobind/portfile.cmake:
```diff
@@ -0,0 +1,25 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO wjakob/nanobind
+    REF 94fa30e22cf2b762a109b7518cd113d2d00dd66a
```
```suggestion
    REF "v${VERSION}"
```

ekilmer committed:
Backport patches instead of waiting for next release

ADGrant commented:
Great news. Thanks.

ekilmer commented:
CI failures don't look related to this PR

dg0yt commented on scripts/ci.baseline.txt:
```diff
@@ -1321,6 +1321,7 @@ vcpkg-ci-mathgl:x64-windows=pass
 vcpkg-ci-mathgl:x64-windows-static=pass
 vcpkg-ci-mathgl:x64-windows-static-md=pass
 vcpkg-ci-mathgl:x86-windows=pass
+vcpkg-ci-nanobind:arm64-windows=fail # cross-compilation failure is expected
```
Why to verify that it fails? The test port is owned by vcpkg. Either we make it succeed, or we disable it via `"supports"` in the manifest.

ekilmer replied:
> Why to verify that it fails?

I was thinking that in case the runner ever changes to a native arm64 machine, then we'll know when it is working, but that's not exactly known (it could still fail while not cross compiling on ARM Windows).

I could try ignoring cross-compilation tests with something like
```cmake
if(VCPKG_CROSSCOMPILING)
    message(WARNING "Skipping vcpkg-ci-nanobind is not expected to work when crosscompiling")
    return()
endif()
```
And then I guess if nanobind ever starts working with cross-compilation, then it'd need to be manually changed.

> we disable it via "supports" in the manifest.

Is it possible to say that cross-compilation is unsupported with `"supports"`? I would think "no" based on the lines in the build system helper port(s)
https://github.com/microsoft/vcpkg/blob/87e969a6d9b388f246f801720bc97d54a7059160/ports/vcpkg-cmake/portfile.cmake#L1-L4

dg0yt replied:
You could say `!native` but this would also block static triplets on the `x64-windows` host.

ekilmer replied:
> You could say `!native`

Thank you. I also see that now in the docs for [platform expressions](https://learn.microsoft.com/en-us/vcpkg/reference/vcpkg-json#platform-expression).

> this would also block static triplets on the `x64-windows` host

Good catch. Using the `if(VCPKG_CROSSCOMPILING)` check is the same as `native`, as the CI reports [here for the static triplet](https://dev.azure.com/vcpkg/public/_build/results?buildId=98304&view=logs&j=3859ca13-59fe-57bd-13f9-449df7d191a6&t=fd8b99e7-4eca-52cb-a612-b91f191f402c&l=7628) 🤔 

It seems like I need a way to compare processor architectures. But we'd need to be able to keep track of compatible runtime architectures like x64 and x86 because a target x86 Python interpreter can run on a host x64 machine.

dg0yt commented on ports/nanobind/usage:
```diff
@@ -0,0 +1,12 @@
+The package nanobind provides CMake functions:
+
+    # Import all Python components potentially needed by nanobind
+    find_package(Python 3.8
+      REQUIRED COMPONENTS Interpreter Development.Module
+      OPTIONAL_COMPONENTS Development.SABIModule)
```
Shouldn't this part be handled by `find_package(nanobind)` (using `find_dependency`?)
Of course it would need to be complemented by dependencies in the manifest. (Does it need an interpreter to run on the host? Is this the reason for the test port to not work for arm64-windows?)

ekilmer replied:
Initially, I kept it this way because I was running into issues somewhere (I forget now). I've pushed a commit for a patch.

ekilmer replied:
> Of course it would need to be complemented by dependencies in the manifest

Are you talking about listing `python` as a dependency in `vcpkg.json` for nanobind? If possible, I would like to merge this PR _without_ doing this.

I want to leave it up to the Python extension writer to choose which Python they use (`vcpkg-ci-nanobind` should prove that it works with vcpkg's Python) so that they can take advantage of vcpkg's ability to manage and track C++ dependencies while still being able to use tools like [`cibuildwheel`](https://github.com/pypa/cibuildwheel) that already set up the necessary Python installation(s).

ekilmer replied:
> Does it need an interpreter to run on the host?

It uses the Python interpreter to figure out certain variables, like the extension suffix, which can be a bit of a pain to hard-code, and the host interpreter would give the wrong result.

ekilmer committed:
Skip vcpkg-ci-nanobind when cross-compiling

ekilmer committed:
Add patch for using `find_dependency` to find Python

ekilmer committed:
Update version

Hoikas commented:
This needs to depend on the python3 port, IMO.

yurybura commented:
Note. The following notes do not apply to your port, but to the implementation of the `nanobind` itself.
"nanobind" does not provide any CMake export targets. As a result, in many cases it is not useful.

1. I can't create a library using my own CMake functions without the `nanobind_add_module` helper function. I would like to use `ninobind`, for example, like other binding libraries:
    ```cmake
    target_link_libraries(mylib
      PRIVATE
        nanobind::link_options
        nanobind::lto
        nanobind::extension_abi3
      PUBLIC
        nanobind::nonobind-abi3
    )
    ```
2. I cannot export and install targets for `mylib` correctly with public dependency to the shared `nanobind` lib:
    ```cmake
    export(TARGETS mylib
        FILE "${PROJECT_BINARY_DIR}/${export_name}.cmake"
    )
    export(PACKAGE ${package_name})
    install(TARGETS mylib
        EXPORT ${export_name}
        RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
        LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
        ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
        PUBLIC_HEADER DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
    )
    ```

ekilmer commented:
@yurybura I agree with your concerns, but it looks like a PR was closed that tries to provide what you say---https://github.com/wjakob/nanobind/pull/201.

There are still cases where this port is useful, like building and packaging a Python native module, where the installation is only expected to be used/loaded by the Python interpreter (and not other CMake projects).

---------------

@Hoikas 

> This needs to depend on the python3 port, IMO.

I can make that change if that's required for getting this merged, but I would argue the primary use case would be for building Python native modules where the majority of users would want Python to come from the system to build against a specific version of Python for their module.

Sure, people could create an empty overlay port for Python, but I feel like that is a bit advanced (but maybe I'm wrong). Maybe we could add some info to the `usage` file? But users would still have to build Python before seeing that text. I'm not sure of the best solution.

ADGrant commented:
@Hoikas I don't see why this port would need to depend on the Python port. I am using pybind11 to wrap a  C++ api via the vcpkg port, I would just like the same support for nanobind. I am not interested in building Python 3 and I don't really want to maintain my own fork of the vcpkg repo with this PR merged.

Hoikas commented:
>  I don't see why this port would need to depend on the Python port. I am using pybind11 to wrap a C++ api via the vcpkg port, I would just like the same support for nanobind. I am not interested in building Python 3 and I don't really want to maintain my own fork of the vcpkg repo with this PR merged.

Your use case is outside the scope of vcpkg. vcpkg intends to control its own ecosystem and not pull from system libraries. It sounds like you're interested in [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html) to include nanbind, IMO. If you really what to build nanobind with vcpkg, you can use an [overlay port](https://learn.microsoft.com/en-us/vcpkg/concepts/overlay-ports) that doesn't depend on python3. Just copy and paste this port into your project and delete the python3 dependency from the manifest. For the main catalog, however, we should provide a solution that just works.

yurybura commented:
@Hoikas The Python3 dependency is redundant for the `nanobond` port since it only copies the source files to the installation directory.

ADGrant commented:
> > I don't see why this port would need to depend on the Python port. I am using pybind11 to wrap a C++ api via the vcpkg port, I would just like the same support for nanobind. I am not interested in building Python 3 and I don't really want to maintain my own fork of the vcpkg repo with this PR merged.
> 
> Your use case is outside the scope of vcpkg. vcpkg intends to control its own ecosystem and not pull from system libraries. It sounds like you're interested in [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html) to include nanbind, IMO. If you really what to build nanobind with vcpkg, you can use an [overlay port](https://learn.microsoft.com/en-us/vcpkg/concepts/overlay-ports) that doesn't depend on python3. Just copy and paste this port into your project and delete the python3 dependency from the manifest. For the main catalog, however, we should provide a solution that just works.

I don't see how my use case is outside the scope of vcpkg. I have a project with a dependency on pybind11, I use vcpkg to consume that dependency without using an overlay port. It just works when I add pybind11 to my vcpkg.json manifest. I would like to be able to switch to nanobind by just changing my manifest to point to nanobind instead of pybind11.   I am not interested in creating an overlay port or building Python 3 from source.  Nor am I interested in using this particular library with another CMake project. The reason I am using nanobind or pybind11 is to provide access to native C++ code from Python.

dg0yt commented:
The binary results must not depend on installation order, or ABI tracking and caching would be broken. Using nanobind after installing port python3 must not behave different from installing just nanobind. Installation order is enforced by ... dependencies.

ebenali commented:
@ekilmer can you re-mark this PR as 'ready' by cycling to draft and back once more to catch the maintainers attention

LilyWangLL commented on ports/nanobind/find_dependency_python.patch:
```diff
@@ -0,0 +1,62 @@
++  set(NB_PYTHON_DEV_MODULE Development.Module)
+ endif()
+ 
++include(CMakeFindDependencyMacro)
++find_dependency(Python 3.8
++  COMPONENTS Interpreter ${NB_PYTHON_DEV_MODULE}
++  OPTIONAL_COMPONENTS Development.SABIModule
++)
```
If you added `find_dependency` here for Python, why not add Python as a dependency, as this will cause usage tests to fail.
```
1> [CMake]   Could NOT find Python (missing: Python_EXECUTABLE Python_LIBRARIES
1> [CMake]   Python_INCLUDE_DIRS Interpreter Development.Module) (Required is at least
1> [CMake]   version "3.8")
```

ekilmer replied:
There's a lot of discussion about this decision in the previous comments on this PR. 

Whatever decision is made here, I believe `nanobind` and `pybind11` should be consistent with whether they list `python3` as a dependency.

ekilmer replied:
> this will cause usage tests to fail

I'm not familiar with these tests. Are they in CI? I could add

```cmake
find_package(Python COMPONENTS Interpreter Development)
```

to the `usage` file like pybind11 does. Would that be better?

LilyWangLL replied:
No, CI does not test usage. And add `find_package(Python COMPONENTS Interpreter Development)` can not resolve this error, because `python3` is not installed.

ekilmer replied:
@LilyWangLL how should I proceed with this?

Is it okay as-is?

Should I add `python3` as a dependency to both nanobind and pybind11?

LilyWangLL replied:
I think you should add `python3` as a dependency to `nanobind`. The PR is to add a new port `nanobind`, so modifications regarding `pybind11` do not need to be included here.

LilyWangLL commented on ports/nanobind/portfile.cmake:
```diff
@@ -0,0 +1,44 @@
+# nanobind distributes source code to build on-demand.
+# The source code is installed into the 'share/${PORT}' directory with
+# subdirectories for source `src` and header `include` files
+set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
```
>  nanobind distributes source code to build on-demand.

Could you provide more detailed explanation about this? Why does the header file of this port need to be installed under `share` directory instead of directly into the `include` directory? I noticed that the path for finding header files in the configuration file is determined by the following code. 
```
get_filename_component(NB_DIR "${CMAKE_CURRENT_LIST_FILE}" PATH)
get_filename_component(NB_DIR "${NB_DIR}" PATH)
```
Perhaps we can adapt the installation strategy of vcpkg port by modifying this?

ekilmer replied:
> > nanobind distributes source code to build on-demand.

> Could you provide more detailed explanation about this?

nanobind's CMake files do not actually build anything. The user of nanobind is responsible for building nanobind's libraries through a set of special CMake functions.

The reason they are installed in `share` is because the paths are hard-coded with relative paths off of a variable holding the root installation path of the source (`NB_DIR`).

I have added a patch. It's rather large, but the changes are simple. I initially avoided doing this because it seems like a maintenance burden (i.e. adding an include file will produce a conflict for the entire thing, and it could be confusing to know which file has been added to make the necessary adjustment), however, it is nice to see nanobind's include files living with all other include files.

LilyWangLL commented on ports/nanobind/portfile.cmake:
```diff
@@ -0,0 +1,44 @@
+# Both of these patches will be included in the next release, and they are
+# required for packaging nanobind
+# https://github.com/wjakob/nanobind/pull/356
+vcpkg_download_distfile(INSTALL_RULES
+    URLS https://github.com/wjakob/nanobind/commit/5bde6527dc43535982a36ffa02d41275c5e484d9.patch?full_index=1
+    SHA512 67bb606483f91bf5dce80881e5ec9f290679c244a745f4c22ef13fd67268cdc81f66f0f1bca9331e567dad06293777a943beaeadf92b7e4e1436e88533daed48
+    FILENAME nanobind-5bde6527dc43535982a36ffa02d41275c5e484d9.patch
+)
+# https://github.com/wjakob/nanobind/pull/359
+vcpkg_download_distfile(MINIMIZE_CMAKE_DEPENDENCIES
+    URLS https://github.com/wjakob/nanobind/commit/978dbb1d6aaeee7530d57cf3e8d558e099a4eec6.patch?full_index=1
+    SHA512 2737235a7aeb111e6dcb4f6d4a96ce7a41f4737a7032a18ebd2a632083849f1a8e48180eda4bdca39d284c0997a546806bcbc7b648cb0c01a9a35f96ba587c8e
+    FILENAME nanobind-978dbb1d6aaeee7530d57cf3e8d558e099a4eec6.patch
+)
```
The latest tag version on upstream is [1.9.2](https://github.com/wjakob/nanobind/releases/tag/v1.9.2), I think this port could update to the new version and remove these patch.

LilyWangLL commented:
I apologize for the delayed review. I have some questions regarding this new port, so I'm changing the status of this PR to draft.

ekilmer committed:
Update nanobind and move include dirs

The diff for moving the include dirs is rather large, but I don't know
of a more concise method

ekilmer committed:
Update version

ekilmer committed:
Add python3 dependency

ekilmer committed:
Update version

BillyONeal commented:
Thanks for the new port!
